### ISSUE TITLE
[Help] Unknown class deserialization

### ISSUE THREAD

**Issue by MelonHell** [Exlll/ConfigLib#14]
i have several classes like Command and Item
```java
public class Command implements Test {
    private String command;

    @Override
    public void run(Player player) {
        Bukkit.dispatchCommand(player, command);
    }
}
```

```java
public class Item implements Test {
    private Material material;
    private int amount;

    @Override
    public void run(Player player) {
        player.getInventory().addItem(new ItemStack(material, amount));
    }
}
```

i want my config to look like this and get a List\<Test\> from it
```yml
myList:
  - type: Item
    material: STONE
    amount: 25
  - type: Command
    command: /say Hello World
```

How can I do this or something similar with this library?

**Comment by Exlll** [comment]
Hi! Using a list that contains subtypes of a configuration class is currently not supported because serializers are selected based on the compile-time type of the list as described in the spoiler of [this section](https://github.com/Exlll/ConfigLib#serializer-selection). What you can do is use a separate list for each type, e.g. `List<Command> commands`, `List<Item> items`, etc.

**Comment by MelonHell** [comment]
> Hi! Using a list that contains subtypes of a configuration class is currently not supported because serializers are selected based on the compile-time type of the list as described in the spoiler of [this section](https://github.com/Exlll/ConfigLib#serializer-selection). What you can do is use a separate list for each type, e.g. `List<Command> commands`, `List<Item> items`, etc.

I really hope that this will appear in one of the future versions, I use this in many of my plugins

**Comment by Exlll** [comment]
I see no reason to not add this functionality. However, the problem is choosing the correct serializers for the individual list items. For example, if you add another command via your configuration file how should the library determine that it belongs to the `Command` and not the `Item` class or even some other subclass of `Test` that happens to have the same fields? Do you have any suggestions?

**Comment by MelonHell** [comment]
> I see no reason to not add this functionality. However, the problem is choosing the correct serializers for the individual list items. For example, if you add another command via your configuration file how should the library determine that it belongs to the `Command` and not the `Item` class or even some other subclass of `Test` that happens to have the same fields? Do you have any suggestions?

I think that I need to somehow indicate in the config which class I want to cast the object to. For example something like this:
```yml
type: Command
command: /say Hello World
```
or this:
```yml
class: com.example.Command
object:
  command: /say Hello World
```

**Comment by Exlll** [comment]
I think that could be a good solution. I'll give it some thought. 

Besides that, may I ask how you plan to use the `List<Test>` list in your code? Determining the type of the items surely involves some kind of `instanceof` checks for you or do you not need to check the types?

**Comment by MelonHell** [comment]
> I think that could be a good solution. I'll give it some thought.
> 
> Besides that, may I ask how you plan to use the `List<Test>` list in your code? Determining the type of the items surely involves some kind of `instanceof` checks for you or do you not need to check the types?

I only need the run(Player player) method from the Test interface from the example at the beginning, but other plugins may need instanceof checks as well.

**Comment by MelonHell** [comment]
In the current plugin, these are csgo-like cases from which various things can drop, such as ItemStack or VIP status or cosmetics from another plugin. I have a CaseItem interface that has a give(Player player) method and I want to add things of different types to the case through the config

**Comment by Exlll** [comment]
So, I've been thinking about this and while handling your specific case (i.e. `List<Test>`) wouldn't be too difficult, I'd prefer having a general solution. The problem becomes quite a bit more complex when dealing with nested collections, like, for example, `List<List<Test>>` (or any other combination). Right now, the `SerializerSelector` doesn't have any context when selecting a serializer: I doesn't know to which field/component a type belongs (and therefore which annotations have been applied to it) nor how deeply nested an element is. To change that, I'll have to rewrite, both, the `SerializerSelector` and `SerializerMapper` classes.

My current idea is to introduce a `SerializeWith` annotation that looks, for example, like this:

```java
@Target({ElementType.FIELD, ElementType.RECORD_COMPONENT})
@Retention(RetentionPolicy.RUNTIME)
public @interface SerializeWith {
    Class<? extends Serializer<?, ?>> serializer();
    int nesting() default 0;
}
```  

Having such an annotation would allow one to apply the referenced serializer at any nesting level:

```java
@SerializeWith(serializer = MyListSerializer.class)
List<List<String>> list0; 

@SerializeWith(serializer = MyListSerializer.class, nesting = 1)
List<List<String>> list1;

@SerializeWith(serializer = MyStringSerializer.class, nesting = 2)
List<List<String>> list2; 
```

Once this functionality has been added, you could easily apply a serializer that does what you want. It would also make it much easier providing yet another annotation that does that automatically.

What do you think about that? Maybe my idea is too complex.

**Comment by MelonHell** [comment]
The @SerializeWith annotation seems to be more convenient than adding a serializer to YamlConfigurationProperties. I don't think it will be difficult

**Comment by Exlll** [comment]
Alright, I've got the `@SerializeWith` annotation working. However, there are still two open questions:

1. How should the serializer whose class is referenced in the annotation be instantiated?

My idea is requiring one of two constructors: Either a no-args constructor or one that takes a `ConfigurationProperties` object and can be used to get access to that object. If both are provided, the latter is called.

2. How should the serialized configurations concretely look like?

While I personally prefer your second suggestion (`class` and `object` field), I don't want to clutter the configuration file hierarchy of users. Hence, I think your first idea is better in this situation. I'd use `==` instead of `type` to designate the type, however. Mostly to prevent name clashes if somebody want to use `type` for a field name but also because Bukkit uses the same name. The config file might then look like this:

```yaml 
list:
- ==: de.exlll.configlib.DynamicTypesSerializerTest$Command
  command: CMD
- ==: de.exlll.configlib.DynamicTypesSerializerTest$Item
  amount: 20
- ==: de.exlll.configlib.DynamicTypesSerializerTest$User
  name: Exlll
```

What do you think?

**Comment by Spliterash** [comment]
You can do it as it is implemented in the jackson library
```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = Command.class, name = "COMMAND"),
        @JsonSubTypes.Type(value = Item.class, name = "ITEM"),
})
```

**Comment by Exlll** [comment]
@Spliterash I don't think that these additional annotations should be required by default. From my perspective, they make configuring quite a bit more complex which I'd like to avoid. Also, adding additional types requires you to go back to all fields that use that type and update the annotations which is another drawback. That said, I still like the flexibility they add.

**Comment by MelonHell** [comment]
> Alright, I've got the `@SerializeWith` annotation working. However, there are still two open questions:
> 
>     1. How should the serializer whose class is referenced in the annotation be instantiated?
> 
> 
> My idea is requiring one of two constructors: Either a no-args constructor or one that takes a `ConfigurationProperties` object and can be used to get access to that object. If both are provided, the latter is called.
> 
>     2. How should the serialized configurations concretely look like?
> 
> 
> While I personally prefer your second suggestion (`class` and `object` field), I don't want to clutter the configuration file hierarchy of users. Hence, I think your first idea is better in this situation. I'd use `==` instead of `type` to designate the type, however. Mostly to prevent name clashes if somebody want to use `type` for a field name but also because Bukkit uses the same name. The config file might then look like this:
> 
> ```yaml
> list:
> - ==: de.exlll.configlib.DynamicTypesSerializerTest$Command
>   command: CMD
> - ==: de.exlll.configlib.DynamicTypesSerializerTest$Item
>   amount: 20
> - ==: de.exlll.configlib.DynamicTypesSerializerTest$User
>   name: Exlll
> ```
> 
> What do you think?

I think it's worth adding the ability to designate aliases for classes in the code, for example, to indicate that Command is de.exlll.configlib.DynamicTypesSerializerTest$Command

**Comment by Exlll** [comment]
Hey, sorry for the late reply, I was on holidays for a week. 

Earier today I added the `@Polymorphic` annotation and pushed it to the `dev` branch for which I had to make quite a few changes to get it working in a way I like (e.g. adding the `SerializeWith` annotation and `SerializerContext` interface). 
I'd appreciate if you could check it out and give me some feedback. It is not yet possible to define type aliases but I'll work on that soon. Also, I haven't updated the README and some other parts of the documentation yet. 

Let me know what you think!

**Usage example:**

```java
package tmp;

import de.exlll.configlib.Configuration;
import de.exlll.configlib.Polymorphic;
import de.exlll.configlib.YamlConfigurations;

import java.nio.file.Paths;
import java.util.List;

public final class Example {
    @Polymorphic
    public interface Action extends Runnable {}

    public record CommandAction(String command) implements Action {
        @Override
        public void run() {
            System.out.println(command);
        }
    }

    @Configuration
    public static class ItemAction implements Action {
        private String material;
        private int amount;

        public ItemAction(String material, int amount) {
            this.material = material;
            this.amount = amount;
        }

        @Override
        public void run() {
            System.out.println(material + ":" + amount);
        }
    }

    public record Config(Action action, List<Action> actions) {}

    public static void main(String[] args) {
        var action = new ItemAction("MATERIAL", 10);
        var actions = List.of(action, new CommandAction("COMMAND"));
        var config = new Config(action, actions);
        YamlConfigurations.save(Paths.get("/tmp/config.yml"), Config.class, config);
    }
}
```

```yaml
action:
  type: tmp.Example$ItemAction
  material: MATERIAL
  amount: 10
actions:
- type: tmp.Example$ItemAction
  material: MATERIAL
  amount: 10
- type: tmp.Example$CommandAction
  command: COMMAND
```

**Comment by Exlll** [comment]
This functionality is now available in version 4.2.0.
